crates/ruff_linter/src/settings/fix_safety_table.rs

```diff
@@ -0,0 +1,166 @@
+use std::fmt::Debug;
+
+use ruff_macros::CacheKey;
+use rustc_hash::FxHashMap;
+use strum::IntoEnumIterator;
+
+use crate::{
+    registry::{Rule, RuleSet},
+    rule_selector::{PreviewOptions, Specificity},
+    RuleSelector,
+};
+
+/// A table to keep track of which rules fixes should have
+/// their safety overridden.
+#[derive(Debug, CacheKey, Default)]
+pub struct FixSafetyTable {
+    forced_safe: RuleSet,
+    forced_unsafe: RuleSet,
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum FixSafety {
+    ForcedSafe,
+    ForcedUnsafe,
+    Default,
+}
+
+impl FixSafetyTable {
+    pub const fn resolve_rule(&self, rule: Rule) -> FixSafety {
+        if self.forced_safe.contains(rule) {
+            FixSafety::ForcedSafe
+        } else if self.forced_unsafe.contains(rule) {
+            FixSafety::ForcedUnsafe
+        } else {
+            FixSafety::Default
+        }
+    }
+
+    pub const fn is_empty(&self) -> bool {
+        self.forced_safe.is_empty() && self.forced_unsafe.is_empty()
+    }
+
+    pub fn from_rule_selectors(
+        extend_safe_fixes: &[RuleSelector],
+        extend_unsafe_fixes: &[RuleSelector],
+        preview_options: &PreviewOptions,
+    ) -> Self {
+        enum Override {
+            Safe,
+            Unsafe,
+        }
+        use Override::{Safe, Unsafe};
```

Comment:
I worry that these names could be confused with our applicability ones — is there a way to do this without defining this temporary enum?

Reply:
We could use boolean. Not a great fan of that as I think it will make this less readable. Your call.